[OpenMP][ASan] Enable ASan Instrumentation for AMDGPUOpenMPToolChain.

ampandey-1995 opened the pull request:
Enable device code ASan instrumentation for openmp offload applications using option '-fsanitize=address'.

jhuber6 reviewed (commented):
Broad question, should we instead aim for having `-fsanitize` work for both sides? I've always been very much against all these `*-gpu` bloat flags when we could just do `-Xarch_host -fsanitize` or `-Xarch_device -fsanitize`. The word against that it usually whether or not it would force people to update some makefiles.

b-sumner commented:
> Broad question, should we instead aim for having `-fsanitize` work for both sides? I've always been very much against all these `*-gpu` bloat flags when we could just do `-Xarch_host -fsanitize` or `-Xarch_device -fsanitize`. The word against that it usually whether or not it would force people to update some makefiles.

I'm not sure I follow the question exactly, but in my opinion it does not make sense to independently control instrumentation for host and device.  We had to implement a means to control instrumentation on the device when development started and after we changed its default once development was complete we kept the option just in case.

jhuber6 commented:
> I'm not sure I follow the question exactly, but in my opinion it does not make sense to independently control instrumentation for host and device. We had to implement a means to control instrumentation on the device when development started and after we changed its default once development was complete we kept the option just in case.

I'm referring  to the fact that for a normal compile, if the user passes `-fsanitize=address`, should that enable it for both the host and device? My argument is that the user is capable of passing something like `-Xarch_host -fsanitize=undefined -Xarch_device -fsanitize=address` if they want fine-grained control, or don't want it on the CPU but want it on the CPU and vice-versa. This patch adds `-fanitize-gpu` which is a completely separate option that only works in this single case.

b-sumner commented:
> I'm referring to the fact that for a normal compile, if the user passes `-fsanitize=address`, should that enable it for both the host and device? My argument is that the user is capable of passing something like `-Xarch_host -fsanitize=undefined -Xarch_device -fsanitize=address` if they want fine-grained control, or don't want it on the CPU but want it on the CPU and vice-versa. This patch adds `-fanitize-gpu` which is a completely separate option that only works in this single case.

I agree, -fsanitize=address is supposed to enable instrumentation on both host and device...and that is exactly what happens now.  If this patch is affecting the existing behavior then something is wrong.

jhuber6 commented on clang/lib/Driver/ToolChains/AMDGPUOpenMP.cpp:
```diff
@@ -37,6 +37,16 @@ AMDGPUOpenMPToolChain::AMDGPUOpenMPToolChain(const Driver &D,
   // discover the 'amdgpu-arch' executable.
   getProgramPaths().push_back(getDriver().Dir);
+  // Diagnose unsupported sanitizer options only once.
+  if (!Args.hasFlag(options::OPT_fgpu_sanitize, options::OPT_fno_gpu_sanitize,
```
@b-sumner I'm basing it on this, but it seems this isn't a new flag, so it's probably some existing behavior?

yxsamliu replied:
-fgpu-sanitize was added during development of sanitizer for AMDGPU. Now it is mainly for debugging purpose to disable sanitizer on device side. Since there are existing apps using it, we need to send a deprecation note in ROCm release before removing it.

b-sumner replied:
Yes.  This is the option to control GPU instrumentation and it is set by default via -fsanitze=address.  Explicitly setting it allows one to disable GPU instrumentation for debugging, etc.

ampandey-1995 replied:
Yes. We actually had a flag ```**-fgpu-sanitize**```  from the beginning of ASan project which actually used by every toolchain(hip,openmp,opencl) to enable device code instrumentation.

Using ```**-fsanitize=address**```  enables asan instrumentation for host code , now for device when ```**-fgpu-sanitize**[Default always on]``` can be used alongwith ```**-fsanitize=address**``` but since it value is by default on so using ```-fgpu-sanitize``` only(without -fsanitize=address) would be  a no-op effect for both host and device code.

We only use the negative ```-fno-gpu-sanitize``` to disable device code asan instrumentation leaving only the host code asan instrumented.

BTW,  other checks are also done in progression i.e in  ```shouldSkipSanitizeFunction``` which checks for ```:xnack+```  feature also in ```--offload-arch=``` option.
 
In short asan instrumentation for both host and device is done using ```-fsanitize=address```  , there are inbuilt clang-driver checks for hip,openmp,opencl which if successful allows -fsanitize=address to pass forward to device toolchain cmd line job invocation otherwise it is skipped.

If any  edge tests cases missing please let me know @b-sumner  , @yxsamliu , @jhuber6 ?

ampandey-1995 commented:
Hi all , the lit test is failing because it is dependent on PR **https://github.com/llvm/llvm-project/pull/123922** . 

Requesting @b-sumner , @yxsamliu , @jhuber6  to rereview again. Sorry in advance!

yxsamliu commented on clang/lib/Driver/ToolChains/AMDGPUOpenMP.cpp:
```diff
@@ -37,6 +37,16 @@ AMDGPUOpenMPToolChain::AMDGPUOpenMPToolChain(const Driver &D,
+    if (K != SanitizerKind::Address)
+      D.getDiags().Report(clang::diag::warn_drv_unsupported_option_for_target)
+          << A->getAsString(Args) << getTriple().str();
+  }
```
HIPAMD toolchain has similar code. Is it possible to refactor them to share common code?

ampandey-1995 replied:
Yes Thanks for reviewing.

ampandey-1995 committed:
Address Comments of @yxsamliu

ampandey-1995 committed:
Fix Windows CI failure.

ampandey-1995 committed:
[OpenMP][ASan] Enable ASan Instrumentation for AMDGPUOpenMPToolChain.

Enable device code ASan instrumentation for openmp offload applications using
option '-fsanitize=address'.

ampandey-1995 committed:
Resolve Merge Conflict.

ampandey-1995 commented:
ping

yxsamliu commented on clang/lib/Driver/ToolChains/AMDGPU.cpp:
```diff
@@ -1106,8 +1106,12 @@ bool AMDGPUToolChain::shouldSkipSanitizeOption(
   // For simplicity, we only allow -fsanitize=address
   SanitizerMask K = parseSanitizerValue(A->getValue(), /*AllowGroups=*/false);
-  if (K != SanitizerKind::Address)
+  if (K != SanitizerKind::Address) {
```
moving diag to this function will cause repeated diagnostic message since the function is called for each argument.

you may want to refactor the diag handling to a common function but keep the call sites at their original locations.

ampandey-1995 replied:
Done.

Hi @yxsamliu & @jhuber6 , there is one complication with respect to OpenMP toolchain i.e ```AMDGPUOpenMPToolchain```.

I am seeing diagnoseUnsupportedSanitizers is getting called twice for AMDGPUOpenMP toolchain.

In ```Driver.cpp```  AMDGPUOpenMPToolchain objects are created twice.

**1. At line number 988.**
`#0  clang::driver::toolchains::ROCMToolChain::diagnoseUnsupportedSanitizers (this=0x7e1ff5c21900, Args=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/ToolChains/AMDGPU.h:154
#1  0x000055555c2e0b53 in clang::driver::toolchains::AMDGPUOpenMPToolChain::AMDGPUOpenMPToolChain (this=0x7e1ff5c21900, D=..., Triple=..., HostTC=..., Args=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/ToolChains/AMDGPUOpenMP.cpp:41                                                                                                                                                                  #2  0x000055555c22489b in std::make_unique<clang::driver::toolchains::AMDGPUOpenMPToolChain, clang::driver::Driver&, llvm::Triple&, clang::driver::ToolChain const&, llvm::opt::InputArgList const&> (__parameter_pack#0=..., __parameter_pack#1=..., __parameter_pack#2=..., __parameter_pack#3=...) at /usr/include/c++/12/bits/unique_ptr.h:1065                                                                                       #3  0x000055555c1fa1ca in clang::driver::Driver::CreateOffloadingDeviceToolChains (this=0x7fffffff8420, C=..., Inputs=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/Driver.cpp:988
#4  0x000055555c2006d6 in clang::driver::Driver::BuildCompilation (this=0x7fffffff8420, ArgList=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/Driver.cpp:1780
#5  0x0000555559685f47 in clang_main(int, char**, llvm::ToolContext const&) ()
#6  0x000055555969a83a in main ()`

**2. At line number 1058**
`#0  clang::driver::toolchains::ROCMToolChain::diagnoseUnsupportedSanitizers (this=0x7e1ff5c23100, Args=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/ToolChains/AMDGPU.h:154
#1  0x000055555c2e0b53 in clang::driver::toolchains::AMDGPUOpenMPToolChain::AMDGPUOpenMPToolChain (this=0x7e1ff5c23100, D=..., Triple=..., HostTC=..., Args=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/ToolChains/AMDGPUOpenMP.cpp:41                                                                                                                                                                  #2  0x000055555c22489b in std::make_unique<clang::driver::toolchains::AMDGPUOpenMPToolChain, clang::driver::Driver&, llvm::Triple&, clang::driver::ToolChain const&, llvm::opt::InputArgList const&> (__parameter_pack#0=..., __parameter_pack#1=..., __parameter_pack#2=..., __parameter_pack#3=...) at /usr/include/c++/12/bits/unique_ptr.h:1065                                                                                       #3  0x000055555c1fad00 in clang::driver::Driver::CreateOffloadingDeviceToolChains (this=0x7fffffff8420, C=..., Inputs=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/Driver.cpp:1058
#4  0x000055555c2006d6 in clang::driver::Driver::BuildCompilation (this=0x7fffffff8420, ArgList=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/Driver.cpp:1780
#5  0x0000555559685f47 in clang_main(int, char**, llvm::ToolContext const&) ()
#6  0x000055555969a83a in main ()`

With Hip there is only one  HIPAMD toolchain.

`#0  clang::driver::toolchains::ROCMToolChain::diagnoseUnsupportedSanitizers (this=0x7e1ff5c21900, Args=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/ToolChains/AMDGPU.h:154
#1  0x000055555c3e4727 in clang::driver::toolchains::HIPAMDToolChain::HIPAMDToolChain (this=0x7e1ff5c21900, D=..., Triple=..., HostTC=..., Args=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/ToolChains/HIPAMD.cpp:220
#2  0x000055555c23278a in std::make_unique<clang::driver::toolchains::HIPAMDToolChain, clang::driver::Driver const&, llvm::Triple const&, clang::driver::ToolChain const&, llvm::opt::ArgList const&> (__parameter_pack#0=..., __parameter_pack#1=..., __parameter_pack#2=..., __parameter_pack#3=...) at /usr/include/c++/12/bits/unique_ptr.h:1065
#3  0x000055555c21df4c in clang::driver::Driver::getOffloadingDeviceToolChain (this=0x7fffffff8400, Args=..., Target=..., HostTC=..., TargetDeviceOffloadKind=@0x7fffffff78f0: clang::driver::Action::OFK_HIP) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/Driver.cpp:6817
#4  0x000055555c1f9b47 in clang::driver::Driver::CreateOffloadingDeviceToolChains (this=0x7fffffff8400, C=..., Inputs=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/Driver.cpp:922
#5  0x000055555c2006d6 in clang::driver::Driver::BuildCompilation (this=0x7fffffff8400, ArgList=...) at /home/ampandey/trunk-toolchain/src-home/llvm-project/clang/lib/Driver/Driver.cpp:1780
#6  0x0000555559685f47 in clang_main(int, char**, llvm::ToolContext const&) ()
#7  0x000055555969a83a in main ()`

ampandey-1995 replied:
Hi @Joseph , 

Does this patch resolves the above issue https://github.com/llvm/llvm-project/pull/125095.

ampandey-1995 replied:
Hi @Joseph , 

Does this patch resolves the above issue https://github.com/llvm/llvm-project/pull/125095.

jhuber6 commented on clang/lib/Driver/ToolChains/AMDGPUOpenMP.cpp:
```diff
@@ -71,10 +71,10 @@ llvm::opt::DerivedArgList *AMDGPUOpenMPToolChain::TranslateArgs(
-  for (Arg *A : Args) {
-    if (!llvm::is_contained(*DAL, A))
+  for (Arg *A : Args)
+    if (!shouldSkipSanitizeOption(*this, Args, BoundArch, A) &&
```
I think this should probably go in `TranslateOpenMPTargetArgs` instead.

ampandey-1995 replied:
Currently I think the  scope of ```shouldSkipSanitizeOption``` is limited to AMDGPU Derived Toolchains only.

ampandey-1995 committed:
Add 'diagnoseUnsupportedSanitizers' function.

Refactor logic of processing unsupported sanitizers per language toolchain to a common 'diagnoseUnsupportedSanitizers' function.

nico commented:
This seems to break tests on my bot: http://45.33.8.238/linux/159455/step_6.txt

Does the error make sense to you?

ampandey-1995 commented:
Fixes https://github.com/llvm/llvm-project/pull/125857
